corehq/motech/auth.py

```diff
@@ -205,16 +206,19 @@ def last_token(self, value: dict):
         self.connection_settings.last_token = value
         self.connection_settings.save()
 
+    @property
+    def should_request_token(self):
+        def token_has_expired(token):
```

Comment:
This will cause unrecoverable errors if the token has been invalidated by the authorization server prior to its expiration and no refresh token is present.

Reply:
> unless we assume a refresh token is always present (which I don't think we should do)?

I think we are talking past each other a bit, because I agree with this assessment. We shouldn't assume a refresh token is present, which is why we shouldn't break repeaters that do not have one. This code will break those repeaters in certain situations (when an access token is invalidated server side before it's expiration date). Those broken repeaters will stay broken until the expiration, which could be days or weeks in the future. That is why I took the, admittedly suboptimal, decision to _always_ request a fresh token if there wasn't a refresh token. We can't know if a token is still valid, so even if it hasn't expired yet, it is safer to request a new one unless we have a way to recover from token validation errors, which we currently do not.

Reply:
Oh, I get it. Yeah, in that case this PR is not adding more value.

Reply:
I'll revert

Reply:
thanks! sorry I raised that in a confusing manner